primitives/src/rule.rs

```diff
@@ -0,0 +1,64 @@
+use crate::{Claim, ClaimType, IdentityId};
+use codec::{Decode, Encode};
+use sp_std::prelude::*;
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+/// It defines the type of rule supported, and the filter information we will use to evaluate as a
+/// predicate.
+pub enum RuleType {
+    /// Rule to ensure that claim filter produces one claim.
+    IsPresent(ClaimType),
+    /// Rule to ensure that claim filter produces an empty list.
+    IsAbsent(ClaimType),
+    /// Rule to ensure that at least one claim is fetched when filter is applied.
+    IsAnyOf(Vec<Claim>),
+    /// Rule to ensure that at none of claims is fetched when filter is applied.
+    IsNoneOf(Vec<Claim>),
+}
+
+impl RuleType {
+    /// It returns the claim type which will be searched and fetched during the evaluation process.
+    /// # NOTE
+    /// The case `IsAnyOf` is special and all the claims should have the same type. The first one
+    /// will be used as the reference type, and any other claim value which differs of that type,
+    /// will be ignored.
+    /// If user defines a empty list of claims in `IsAnyOf`, `Jurisdiction` type will be used by
+    /// default.
+    pub fn as_claim_type(&self) -> ClaimType {
+        match self {
+            RuleType::IsPresent(claim_type) => *claim_type,
+            RuleType::IsAbsent(claim_type) => *claim_type,
+            RuleType::IsNoneOf(ref claims) => Self::get_claim_type(claims.as_slice()),
+            RuleType::IsAnyOf(ref claims) => Self::get_claim_type(claims.as_slice()),
+        }
+    }
+
+    fn get_claim_type(claims: &[Claim]) -> ClaimType {
+        claims
+            .iter()
+            .map(|claim| claim.claim_type())
+            .nth(0)
+            .unwrap_or(ClaimType::Jurisdiction)
+    }
+}
+
+/// Type of claim requirements that a rule can have
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct Rule {
+    /// Type of rule.
+    pub rule_type: RuleType,
+    /// Trusted issuers.
+    pub issuers: Vec<IdentityId>,
+    /// Claim scope
+    pub scope: Option<IdentityId>,
```

Comment:
I think scope should be part of the `rule_type` so that people can do something like "IsAnyOf whitelist with scope A or whitelist with scope B".

Reply:
That is a pretty nice use case. However if we add `scope` into `rule_type`, the evaluation of non-scoped claims (like CDD) would be more complex (at least on my mind).
My suggestion is to extend Rule to support AND, OR operations (similar to Predicate) which will be more expressive and flexible.
So for instance:
```
 Rule { 
    rule_type: IsAnyOf ( x, y, z),
    scope: 'A' 
}. or( Rule {
    rule_type: IsAnyOf( x', y', z'),
    scope: 'B'
}). and( Rule {
    rule_type: IsPresent(CDD)
}
```

BTW in order to limit this task, we can evaluate that extensions/ideas on another task.

Reply:
I think this is one of the main product requirements and needs to be implemented. That being said, I agree that it might make sense to do in a different task. cc @adamdossa WDYT?